### PR TITLE
Refactor: Move SNS Aggregator types and converters

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Clean up the SNS Aggregator types before refactoring the api and services.

# Changes

* Move types and converters of the SNS Aggregator to separate files.

# Tests

* Add a test for the main converter.

# Todos

Not worth an entry, only refactor. The interface of the sns aggregator api is exactly the same.

### CODE REVIEW

File: frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { convertDtoData } from "$lib/utils/sns-aggregator-converters.utils";
+import {
+  aggregatorSnsMock,
+  aggregatorSnsMockDto,
+} from "$tests/mocks/sns-aggregator.mock";
+
+describe("sns aggregator converters utils", () => {
+  describe("convertDtoData", () => {
+    // This might not test much, but it allowed me to catch some mismatch in the types
```
Comment (dskloetd) [inline frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts:9]: Does that mean that the types are not a direct copy and also contain some fixes? It's impossible to see the way GitHub displays it but would have been nice to see.

File: frontend/src/tests/mocks/sns-aggregator.mock.ts [hunk 2]
```diff
@@ -295,10 +309,14 @@ export const aggregatorSnsMockWith = ({
     root_canister_id: rootCanisterId,
   },
   swap_state: {
-    ...aggregatorSnsMock.swap_state,
```
Comment (dskloetd) [inline frontend/src/tests/mocks/sns-aggregator.mock.ts:298]: Why did you remove this?

Comment (lmuntaner) [reply frontend/src/tests/mocks/sns-aggregator.mock.ts:298]: Yes, I'll revert. Not needed.

File: frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts:9]: Yes, sorry. I had to do a small change but I didn't think it was worth a separate PR for this.

Before
```
const convertIcrc1Metadata = (
  icrc1Metadata: CachedSnsTokenMetadataDto
): IcrcTokenMetadataResponse => {
  return icrc1Metadata.map(([key, value]) => {
    if ("Int" in value) {
      return [key, { Int: BigInt(value.Int[0]) }];
    }
    if ("Nat" in value) {
      return [key, { Nat: BigInt(value.Nat[0]) }];
    }
    return [key, value];
  });
};
```

Now:
```
const convertIcrc1Metadata = (
  icrc1Metadata: CachedSnsTokenMetadataDto
): IcrcTokenMetadataResponse => {
  return icrc1Metadata.map(([key, value]) => {
    if ("Int" in value) {
      return [key, { Int: BigInt(value.Int[0]) }];
    }
    if ("Nat" in value) {
      return [key, { Nat: BigInt(value.Nat[0]) }];
    }
    if ("Text" in value) {
      return [key, { Text: value.Text }];
    }
    if ("Nat64" in value) {
      return [key, { Nat: BigInt(value.Nat64[0]) }];
    }
    // We don't need nor support Blob, Array and Map yet
    return [key, { Text: "unsopported" }];
  });
};
```

File: frontend/src/tests/mocks/sns-aggregator.mock.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/tests/mocks/sns-aggregator.mock.ts]: done

File: frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts:9]: The type also changed:

Before:
```
type CachedSnsTokenMetadataDto = [
  string | IcrcMetadataResponseEntries,
  (
    | { Int: [number] }
    | { Nat: [number] }
    | { Blob: Uint8Array }
    | { Text: string }
  )
][];
```

Now:
```
type MapDto = Array<[string, ValueDto]>;

export type ValueDto =
  | { Int: [number] }
  | { Map: MapDto }
  | { Nat: [number] }
  | { Nat64: [number] }
  | { Blob: number[] }
  | { Text: string }
  | { Array: Array<ValueDto> };

export type CachedSnsTokenMetadataDto = [
  string | IcrcMetadataResponseEntries,
  ValueDto
][];
```

File: frontend/src/lib/utils/sns-aggregator-converters.utils.ts [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+      return [key, { Nat: BigInt(value.Nat[0]) }];
+    }
+    if ("Text" in value) {
+      return [key, { Text: value.Text }];
+    }
+    if ("Nat64" in value) {
+      return [key, { Nat: BigInt(value.Nat64[0]) }];
+    }
+    // We don't need nor support Blob, Array and Map yet
+    return [key, { Text: "unsopported" }];
```
Comment (dskloetd) [inline frontend/src/lib/utils/sns-aggregator-converters.utils.ts:245]: Can we throw an error here or would that be too little "support"?

File: frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts:9]: Sorry, I also forgot a new converter:

```
const convertFunctionType = (
  functionType: CachedFunctionTypeDto
): SnsFunctionType => {
  if ("NativeNervousSystemFunction" in functionType) {
    return { NativeNervousSystemFunction: {} };
  }
  const { GenericNervousSystemFunction } = functionType;
  return {
    GenericNervousSystemFunction: {
      validator_canister_id: convertOptionalStringToOptionalPrincipal(
        GenericNervousSystemFunction.validator_canister_id
      ),
      target_canister_id: convertOptionalStringToOptionalPrincipal(
        GenericNervousSystemFunction.target_canister_id
      ),
      validator_method_name: toNullable(
        GenericNervousSystemFunction.validator_method_name
      ),
      target_method_name: toNullable(
        GenericNervousSystemFunction.target_method_name
      ),
    },
  };
};
```

File: frontend/src/lib/utils/sns-aggregator-converters.utils.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/lib/utils/sns-aggregator-converters.utils.ts:245]: I'd not throw here. Tokens might have more metadata than we use. Therefore, I'd rather not throw just because a token has more metadata than we use.

This will just map the unsupported metadata to this `Text` entry.

File: frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts:9]: Sorry for not doing it in another PR... I was a little bit lazy to do two separate PRs for these 🙈

**Comment by lmuntaner** [comment]
Closing this one in favor of #3034

File: frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/sns-aggregator-converters.utils.spec.ts:9]: New PR: https://github.com/dfinity/nns-dapp/pull/3034
